dbms/src/Flash/Mpp/MPPTaskId.h

```diff
@@ -56,20 +56,34 @@ struct MPPQueryIdHash
 };
 
 /// A MPP query has one or more MPPGathers, each mpp gather has one or more MPPTasks. The mpp tasks in different mpp gathers are independent
-/// to each other, so theoretically speaking, the smallest cancel/retry unit of MPP query could be MPPGather. However, MPPGathers belong to
-/// the same MPP query could have dependence to each other(e.g. for query A join B, if the join is not supported in TiFlash, TiDB will generate
-/// two mpp gathers, one is reading from A and the other is reading from B, the probe side's mpp gather depends on the build side's mpp gather),
-/// so the smallest scheduling unit in TiFlash is still the MPP query
-/// Note currently, TiDB does not fill `gather_id` in mpp::TaskMeta, TiFlash hard coded 0 as the gather id for all MPPGatherId, so a mpp query now
-/// only has one MPPGather, and the schedule/cancel/retry unit in TiFlash is MPP query, we may implement mpp gather level's cancel/retry in the future.
+/// to each other, while MPPGathers belong to the same MPP query could have dependence to each other(e.g. for query A join B, if the join is
+/// not supported in TiFlash, TiDB will generate two mpp gathers, one is reading from A and the other is reading from B, the probe side's mpp
+/// gather depends on the build side's mpp gather), so the smallest scheduling unit in TiFlash is MPP query, but the smallest cancel/retry unit
+/// in TiFlash is MPP gather.
 struct MPPGatherId
 {
-    UInt64 gather_id;
+    Int64 gather_id;
     MPPQueryId query_id;
     MPPGatherId(Int64 gather_id_, const MPPQueryId & query_id_)
         : gather_id(gather_id_)
         , query_id(query_id_)
     {}
+    MPPGatherId(Int64 gather_id_, UInt64 query_ts, UInt64 local_query_id, UInt64 server_id, UInt64 start_ts)
+        : gather_id(gather_id_)
+        , query_id(query_ts, local_query_id, server_id, start_ts)
+    {}
+    explicit MPPGatherId(const mpp::TaskMeta & task_meta)
+        : gather_id(task_meta.gather_id())
+        , query_id(task_meta)
+    {}
+    String toString() const
+    {
+        return fmt::format("<gather_id:{}, query_ts:{}, local_query_id:{}, server_id:{}, start_ts:{}>", gather_id, query_id.query_ts, query_id.local_query_id, query_id.server_id, query_id.start_ts);
+    }
+    bool hasMeaningfulGatherId() const
+    {
+        return gather_id > 0;
```

Comment:
Does the tidb side guarantee that the gather id must be greater than 0?

Reply:
After https://github.com/pingcap/tidb/pull/45226, it is expected that all the gather_id in `TaskMeta` is greater than 0, before that pr, it is not guaranteed. In TiFlash side, I add some checks to ensure that if a mpp query has a meaningful gather id(greater that 0), then all the tasks belongs to the same mpp query must all have a meaningful gather id. That is to say, TiFlash only support two cases:
- All the mpp tasks have a meaningfull gather id
- All the mpp tasks have 0 as gather id
And luckily there is no compatibility issues since for TiDB version <= v7.2.0, all mpp tasks have 0 as the gather id, and for TiDB version >= 7.3.0, all mpp tasks should have a meaningful gather id.

Reply:
ok